concordium-consensus/src/Concordium/KonsensusV1/Consensus/Finality.hs

```diff
@@ -48,23 +50,45 @@ checkFinality ::
       MonadThrow m,
       MonadConsensusEvent m,
       MonadLogger m,
-      IsConsensusV1 (MPV m)
+      IsConsensusV1 (MPV m),
+      HasCallStack
     ) =>
-    QuorumCertificate ->
+    -- |The newly-certified block.
+    CertifiedBlock (MPV m) ->
     m ()
-checkFinality qc = do
-    sd <- get
-    case getMemoryBlockStatus (qcBlock qc) sd of
-        Just (BlockAlive block) -> checkFinalityWithBlock qc block
-        _ -> return ()
+processCertifiedBlock cb@CertifiedBlock{..}
+    | NormalBlock block <- bpBlock cbQuorumBlock,
+      let parentQC = blockQuorumCertificate block,
+      qcRound cbQuorumCertificate == qcRound parentQC + 1,
+      qcEpoch cbQuorumCertificate == qcEpoch parentQC = unlessStored $ do
+        let finalizedBlockHash = qcBlock parentQC
+        sd <- get
+        unless (finalizedBlockHash == getHash (sd ^. lastFinalized)) $ do
+            let !newFinalizedPtr = parentOfLive sd cbQuorumBlock
+            let newFinalizationEntry =
+                    FinalizationEntry
+                        { feFinalizedQuorumCertificate = parentQC,
+                          feSuccessorQuorumCertificate = cbQuorumCertificate,
+                          feSuccessorProof = getHash (sbBlock block)
+                        }
+            processFinalizationHelper newFinalizedPtr newFinalizationEntry (Just cb)
+            shrinkTimeout cbQuorumBlock
+    | otherwise = unlessStored $ do
+        storedBlock <- makeStoredBlock cbQuorumBlock
+        LowLevel.writeCertifiedBlock storedBlock cbQuorumCertificate
+  where
+    unlessStored a = do
```

Comment:
Please document that this is to cover the case where we have rolled back blocks (and we don't "demote" a finalized block as part of this process).